### PR TITLE
Add error for trailing angle brackets.

### LINKED ISSUE #54521: Misleading error message with malformed turbofishes
I typoed the angle brackets when attempting turbofish.

```rust
fn main() {
    println!("{:?}", vec![1, 2, 3].into_iter().collect::<Vec<usize>>>());
}

```

([Playground](https://play.rust-lang.org/?gist=dd291f4a7c8e184f571f8d00f8dc839c&version=nightly&mode=debug&edition=2015))

I expected to see this happen: To get an error indicating that I had an extra closing
angle bracket after my turbofish, or that I had a mismatched delimiter.

Here is what happened instead:

Errors:

```
   Compiling playground v0.0.1 (/playground)
error: field expressions may not have generic arguments
 --> src/lib.rs:2:55
  |
2 |     println!("{:?}", vec![1, 2, 3].into_iter().collect::<Vec<usize>>>());
  |                                                       ^^^^^^^^^^^^^^

error: aborting due to previous error

error: Could not compile `playground`.

To learn more, run the command again with --verbose.

```

This error message refers to the case when you attempt to use turbofish on a struct field
rather than a method or function, like `struct.field::<usize>`. However, this error message
is confusing to users because, if you look at the end of the collect call, there _are_ parentheses,
and this _looks_ like a correctly formed method call, which obscures the real error.

### PR DESCRIPTION
**Pull request by davidtwco**
Fixes #54521.

This PR adds a error (and accompanying machine applicable
suggestion) for trailing angle brackets on function calls with a
turbofish.

r? @estebank

### CODE REVIEW

**Comment by davidtwco** [comment]
> 
> 
> I think you can change this to use `any` instead of `map().unwrap_or()`.

I might be missing something, but I don't think I can because this is a `Option<P<GenericArgs>>`?

**Comment by estebank** [comment]
> I might be missing something, but I don't think I can because this is a Option<P<GenericArgs>>?

Ah, you're right. I thought it was something like `Vec<P<GenericArg>>`. Carry on.

**Comment by estebank** [comment]
r=me once travis finishes

File: src/libsyntax/parse/parser.rs [hunk 1]
```diff
@@ -2149,7 +2149,27 @@ impl<'a> Parser<'a> {
...
+                //
+                // As opposed to the below highlight (if we had only finished the first
+                // recursion):
+                //
+                // `Foo::<Bar as Baz<T>>::Qux`
+                //                     ^ here
+                //
+                // `PathStyle::Expr` is only provided at the root invocation and never in
+                // `parse_path_segment` to recurse and therefore can be checked to maintain
+                // this invariant.
```
Comment (davidtwco) [inline src/libsyntax/parse/parser.rs]: @estebank do you know if this last sentence is accurate? It works and passes tests but I'm not sure if `PathStyle::Expr` was intended to make this condition hold.

**Review by davidtwco — COMMENTED** [review]
@estebank I've added a new commit handling another case, would appreciate if you could re-review.

File: src/libsyntax/parse/parser.rs (hunk 1 above)
Comment (estebank) [reply src/libsyntax/parse/parser.rs]: It should be ok.

**Pull request merged**